In [1]:
!pip install tabulate

In [2]:
# Required imports
%matplotlib inline
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import warnings
warnings.filterwarnings('ignore')

import yfinance as yf

# Set ticker and date range
ticker = 'TSLA'
start_date = '2014-01-01'
end_date = '2024-12-31'

# Download data directly using yfinance
df = yf.download(ticker, start=start_date, end=end_date)

# Extract only 'Close' column
tsla_close = df['Close'].copy()

# Display sample data
#print(tsla_close.head())



YF.download() has changed argument auto_adjust default to True


[*********************100%***********************]  1 of 1 completed


In [9]:
tsla_close.head(3)

Ticker           TSLA
Date                 
2014-01-02  10.006667
2014-01-03   9.970667
2014-01-06   9.800000

In [ ]:
# Save to CSV
#tsla_close.to_csv('tesla_close_2014_2024.csv')

#print("✅ Tesla Close price data (2014–2024) saved as 'tesla_close_2014_2024.csv'")


In [13]:
import yfinance as yf

# Load Tesla ticker
tsla = yf.Ticker("TSLA")

# Get financial statements
income_stmt = tsla.financials
balance_sheet = tsla.balance_sheet
cashflow_stmt = tsla.cashflow

# Transpose for year-wise view
income_stmt = income_stmt.T
balance_sheet = balance_sheet.T
cashflow_stmt = cashflow_stmt.T

# Filter 2022 to 2024
years = ['2022-12-31', '2023-12-31', '2024-12-31']
income_stmt_filtered = income_stmt.loc[income_stmt.index.astype(str).isin(years)]
balance_sheet_filtered = balance_sheet.loc[balance_sheet.index.astype(str).isin(years)]
cashflow_stmt_filtered = cashflow_stmt.loc[cashflow_stmt.index.astype(str).isin(years)]

# Display samples
# print("Income Statement (2022–2024):\n", income_stmt_filtered)
# print("\nBalance Sheet (2022–2024):\n", balance_sheet_filtered)
# print("\nCash Flow Statement (2022–2024):\n", cashflow_stmt_filtered)



In [ ]:
# # Save to CSV files
# income_stmt_filtered.to_csv("tesla_income_statement_2022_2024.csv")
# balance_sheet_filtered.to_csv("tesla_balance_sheet_2022_2024.csv")
# cashflow_stmt_filtered.to_csv("tesla_cashflow_statement_2022_2024.csv")


S&P 500 index (^GSPC) historical “Close” prices from 2014 to 2024

In [10]:
import yfinance as yf
import pandas as pd

# Define ticker and date range
ticker = '^GSPC'  # S&P 500 index
start_date = '2014-01-01'
end_date = '2024-12-31'

# Download the data
sp500_data = yf.download(ticker, start=start_date, end=end_date)

# Extract only the 'Close' column
sp500_close = sp500_data[['Close']].copy()




[*********************100%***********************]  1 of 1 completed


In [11]:
# # Save to CSV
# sp500_close.to_csv('sp500_close_2014_2024.csv')

# print("✅ S&P 500 Close price data (2014–2024) saved as 'sp500_close_2014_2024.csv'")

✅ S&P 500 Close price data (2014–2024) saved as 'sp500_close_2014_2024.csv'


#### Liquidity Ratios

In [ ]:
current ratio

In [16]:
# Transpose so that line items are rows and dates are columns
balance_sheet_filtered_T = balance_sheet_filtered.T

# Extract current assets and liabilities
current_assets = balance_sheet_filtered_T.loc['Current Assets']
current_liabilities = balance_sheet_filtered_T.loc['Current Liabilities']

# Calculate the current ratio
current_ratio = current_assets / current_liabilities
current_ratio.name = "Current Ratio"

# Display the result
current_ratio


2024-12-31    2.024912
2023-12-31    1.725894
2022-12-31    1.531956
Name: Current Ratio, dtype: object

A current ratio of 2.02 on 2024-12-31 means:

For every $1 of short-term liabilities due within a year, Tesla had $2.02 in short-term assets on hand.
Financially healthy: A ratio above 1 indicates Tesla can cover its short-term obligations without stress.

It suggests Tesla has twice as many current assets as current liabilities—a solid liquidity cushion.

In [ ]:
Quick Ratio

In [18]:
# Extract required values
current_assets = balance_sheet_filtered_T.loc['Current Assets']
inventory = balance_sheet_filtered_T.loc['Inventory']
current_liabilities = balance_sheet_filtered_T.loc['Current Liabilities']

# Calculate Quick Ratio
quick_ratio = (current_assets - inventory) / current_liabilities
quick_ratio.name = "Quick Ratio"

# Display result
quick_ratio


2024-12-31    1.607959
2023-12-31    1.251913
2022-12-31    1.051256
Name: Quick Ratio, dtype: object

For every $1 of Tesla’s short-term liabilities in 2024, it had $1.61 in liquid assets—excluding inventory—to cover them immediately.

In [ ]:
Cash Ratio

In [19]:
# For Cash Ratio: (Cash + Cash Equivalents) / Current Liabilities

# Extract required components
cash = balance_sheet_filtered_T.loc['Cash And Cash Equivalents']
current_liabilities = balance_sheet_filtered_T.loc['Current Liabilities']

# Calculate Cash Ratio
cash_ratio = cash / current_liabilities
cash_ratio.name = "Cash Ratio"

# Display result
cash_ratio


2024-12-31    0.559974
2023-12-31    0.570405
2022-12-31    0.608521
Name: Cash Ratio, dtype: object

A ratio around 0.5–0.6  indicates a strong liquidity position
On 2024-12-31, Tesla had $0.56 in pure cash and cash equivalents for every $1 of short-term liabilities.

#### Profitability Ratios

Gross Profit Margin(from income sheet)

In [22]:
# Transpose to make line items rows and dates columns
income_stmt_T = income_stmt.T

# Extract required components
revenue = income_stmt_T.loc['Total Revenue']
gross_profit = income_stmt_T.loc['Gross Profit']

# Gross Profit Margin = Gross Profit/Net Sales * 100

# Calculate Gross Profit Margin = Gross Profit / Revenue
gross_profit_margin = gross_profit / revenue
gross_profit_margin.name = "Gross Profit Margin"

# Display result
gross_profit_margin


2024-12-31    0.178626
2023-12-31    0.182489
2022-12-31    0.255984
2021-12-31    0.252792
2020-12-31         NaN
Name: Gross Profit Margin, dtype: object

Gross Profit Margin in year 2024 is 17.8%

A clear decline in gross margin from 2022 to 2024.

In 2022, Tesla kept 26 cents of every $1 in revenue after covering cost of goods.

By 2024, that dropped to 18 cents—suggesting:

Rising production costs (e.g., batteries, logistics)

Price cuts or competitive pressure



Operating Profit Margin

In [23]:
# Extract required components from income statement
operating_income = income_stmt_T.loc['Operating Income']
total_revenue = income_stmt_T.loc['Total Revenue']

# Calculate Operating Profit Margin = Operating Income / Total Revenue
operating_profit_margin = operating_income / total_revenue
operating_profit_margin.name = "Operating Profit Margin"

# Display result
operating_profit_margin


2024-12-31    0.079435
2023-12-31    0.091875
2022-12-31    0.169797
2021-12-31    0.120692
2020-12-31         NaN
Name: Operating Profit Margin, dtype: object

Here is Tesla’s Operating Profit Margin for each year:

2022-12-31 → 16.98%

2023-12-31 → 9.19%

2024-12-31 → 7.94%

This margin drop aligns with the earlier gross margin decline, suggesting Tesla faced cost pressures not offset by revenue growth or cost controls.

Return on Assets (ROA)

In [24]:
# Load total assets from the balance sheet
total_assets = balance_sheet_filtered_T.loc['Total Assets']

# Load net income from the income statement
net_income = income_stmt_T.loc['Net Income']

# Calculate Return on Assets (ROA) = Net Income / Total Assets
roa = net_income / total_assets
roa.name = "Return on Assets (ROA)"

# Display result
roa


2020-12-31         NaN
2021-12-31         NaN
2022-12-31    0.152821
2023-12-31     0.14068
2024-12-31    0.058409
Freq: A-DEC, Name: Return on Assets (ROA), dtype: object

Here is Tesla’s Return on Assets (ROA) for each year:

2022-12-31 → 15.28%

2023-12-31 → 14.07%

2024-12-31 → 5.84%

Return on Equity (ROE)

In [25]:
# Load shareholders' equity from the balance sheet
shareholders_equity = balance_sheet_filtered_T.loc['Total Equity Gross Minority Interest']

# Use previously loaded net income
# Calculate Return on Equity (ROE) = Net Income / Shareholders' Equity
roe = net_income / shareholders_equity
roe.name = "Return on Equity (ROE)"

# Display result
roe


2020-12-31         NaN
2021-12-31         NaN
2022-12-31    0.274151
2023-12-31      0.2358
2024-12-31     0.09677
Freq: A-DEC, Name: Return on Equity (ROE), dtype: object

2022-12-31 → 27.42%

2023-12-31 → 23.58%

2024-12-31 → 9.68%

ROE shows how much profit Tesla generated for each dollar of shareholder equity
By 2024, ROE dropped sharply, mirroring the decline in ROA

Return on sales (ROS)

In [28]:
##froom income statement
# ROS = Operating Income / Total Revenue
ros = operating_income / total_revenue
ros.name = "Return on Sales "

# Display the correct ROS
ros


2024-12-31    0.079435
2023-12-31    0.091875
2022-12-31    0.169797
2021-12-31    0.120692
2020-12-31         NaN
Name: Return on Sales , dtype: object

2022-12-31 → 16.98%

2023-12-31 → 9.19%

2024-12-31 → 7.94%

A Return on Sales (ROS) of 7.94% on 2024-12-31 means:

For every $1 of revenue Tesla generated in 2024, it earned 7.94 cents in operating profit after covering all operating costs—but before interest and taxes

Return on Investment (ROI)

In [29]:
# Load total debt and shareholders' equity from balance sheet
total_debt = balance_sheet_filtered_T.loc['Total Debt']
shareholders_equity = balance_sheet_filtered_T.loc['Total Equity Gross Minority Interest']

# Compute invested capital
invested_capital = total_debt + shareholders_equity

#ROI = Investment Gain / Investment Base
# Compute ROI = Net Income / Invested Capital
roi = net_income / invested_capital
roi.name = "Return on Investment (ROI)"

# Display result
roi


2020-12-31         NaN
2021-12-31         NaN
2022-12-31    0.243639
2023-12-31    0.204955
2024-12-31     0.08167
Freq: A-DEC, Name: Return on Investment (ROI), dtype: object

Here is Tesla’s Return on Investment (ROI) using the formula Net Income / (Total Debt + Shareholders' Equity):

2022-12-31 → 24.36%

2023-12-31 → 20.50%

2024-12-31 → 8.17%

A steep drop in 2024 suggests:

Lower net income

Possibly rising invested capital (new facilities, tech investment)

#### Earning Ratio

In [35]:
tsla_close_before = tsla_close[:'2024-12-30'].tail(1)
print(tsla_close_before)


Ticker            TSLA
Date                  
2024-12-30  417.410004


In [37]:
# EPS calculation
income_stmt_T = income_stmt.T
balance_sheet_T = balance_sheet.T

net_income = income_stmt_T.loc['Net Income']
shares_outstanding = balance_sheet_T.loc['Ordinary Shares Number']
eps = net_income / shares_outstanding
eps.name = "Earnings Per Share"

# Get closing price on or before 2024-12-30 as scalar
tsla_price = tsla_close[:'2024-12-30'].tail(1).values[0]

# Compute P/E ratio: scalar divided by Series
pe_ratio = tsla_price / eps
pe_ratio.name = "Price-to-Earnings (P/E) Ratio"

# Display result
print(pe_ratio)


2024-12-31    188.273572
2023-12-31     88.635967
2022-12-31      104.9579
2021-12-31    234.169732
2020-12-31           NaN
Name: Price-to-Earnings (P/E) Ratio, dtype: object


Investors were paying $188.27 for every $1 of Tesla’s earnings in 2024. 

Debt-to-Equity Ratio

In [39]:
# Extract Total Debt and Total Equity from the balance sheet
total_debt = balance_sheet_T.loc['Total Debt']
total_equity = balance_sheet_T.loc['Total Equity Gross Minority Interest']

# Calculate Debt-to-Equity Ratio = Total Debt / Total Equity
debt_to_equity_ratio = total_debt / total_equity
debt_to_equity_ratio.name = "Debt-to-Equity Ratio"

# Display result
debt_to_equity_ratio


2024-12-31    0.184894
2023-12-31    0.150498
2022-12-31    0.125234
2021-12-31    0.280942
2020-12-31         NaN
Name: Debt-to-Equity Ratio, dtype: object

This ratio indicates how much debt Tesla is using to finance its assets relative to equity.

In 2024, Tesla had $0.18 of debt for every $1 of equity.

#### Sustainable Growth Rate

In [ ]:
Sustainable Growth Rate Formula:
SGR=ROE×(1−Dividend Payout Ratio)
Where:

ROE = Return on Equity

Dividend Payout Ratio = Dividends / Net Income



according to Yahoo Finance, Tesla does not pay dividends. The "Key Statistics" page for Tesla (TSLA) indicates:​
Yahoo Finance

Trailing Annual Dividend Rate: 0.00​

Trailing Annual Dividend Yield: 0.00%​

5 Year Average Dividend Yield: --​

These figures confirm that Tesla has not issued dividends to its shareholders

In [40]:
# Since Tesla pays no dividends, the retention ratio = 1
# Thus, Sustainable Growth Rate = ROE

# Load previously computed ROE
roe = net_income / total_equity
roe.name = "Sustainable Growth Rate (SGR)"

# Display the result
roe


2024-12-31     0.09677
2023-12-31      0.2358
2022-12-31    0.274151
2021-12-31    0.174904
2020-12-31         NaN
Name: Sustainable Growth Rate (SGR), dtype: object

2024-12-31 → 9.68%